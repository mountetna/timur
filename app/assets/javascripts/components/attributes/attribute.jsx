import { reviseDocument } from '../../actions/magma_actions'

var Attribute = React.createClass({
  render: function() {
    var self = this
    var store = this.context.store
    if (this.props.mode == "edit") {
      return <div className="value">
              <SlowTextInput textClassName="full_text" 
                waitTime={500}
                placeholder={ this.props.attribute.placeholder }
                update={
                  function(value) {
                    store.dispatch(reviseDocument(
                      self.props.document,
                      self.props.template,
                      self.props.attribute,
                      value)
                    )
                  }
                }
                defaultValue={ this.props.revision } />
             </div>
    }

    return <div className="value">
            { this.props.value }
           </div>
  }
})
Attribute.contextTypes = {
  store: React.PropTypes.object
}

module.exports = Attribute;
