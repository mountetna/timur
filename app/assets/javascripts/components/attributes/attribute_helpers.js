
var AttributeHelpers = {
  attribute_exists: function() {
    return(this.attribute_value() != null)
  },
  attribute_value: function() {
    return(this.props.record[this.props.attribute.name]);
  },
  value_name: function() {
    return "values[" + this.props.attribute.name + "]";
  },
  link_name: function() {
    return "link[" + this.props.attribute.name + "]";
  },
  unlink_name: function() {
    return "unlink[" + this.props.attribute.name + "]";
  }
};

module.exports = AttributeHelpers;
