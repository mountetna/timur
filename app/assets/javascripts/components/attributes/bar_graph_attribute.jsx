import React from 'react'
import BarGraph from '../plots/bar_graph'
import { selectConsignment } from '../../selectors/consignment'

let BarGraphAttribute = ({
  data,
  datumKey,
  ymin,
  ymax,
  attribute: {
    plot: {
      name,
      dimensions: {
        height,
        width,
        margin
      }
    }
  }

}) => (
  <div className="value">
    {data[0] &&
      <BarGraph
        ymin={ymin}
        ymax={ymax}
        plot={{name, width, height}}
        margin={margin}
        data={data}
      />
    }
  </div>
)

BarGraphAttribute = connect(
  (state, props) => {
    const { name } = props.attribute.plot
    const consignment = selectConsignment(state, name)
    if (consignment) {
      const data = consignment.data.map((label, value,) => {
        return value.map((label, value) => ({ label, value })).reduce((acc, curr) => {
          return {
            ...acc,
            [curr.label]: curr.value
          }
        }, {})
      })

      return { data }
    }

      return {
        data: []
      }
  }
)(BarGraphAttribute)

export default BarGraphAttribute
