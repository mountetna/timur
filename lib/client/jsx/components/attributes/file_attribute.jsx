// Framework libraries.
import React, {useState, useEffect, useRef} from 'react';
import {connect} from 'react-redux';
import Modal from 'react-modal';

import {useReduxState} from 'etna-js/hooks/useReduxState';
import {useActionInvoker} from 'etna-js/hooks/useActionInvoker';

import {
  reviseDocument,
  getRevisionTempUrl,
  fetchTempUrl
} from '../../actions/magma_actions';
import ButtonBar from '../button_bar';
import Icon from '../icon';

export const STUB = '::blank';
export const TEMP = '::temp';

// metis:\/\/([^\/]*?)\/([^\/]*?)\/(.*)$
const METIS_PATH_MATCH = (path) =>
  new RegExp(
    '^metis://' +
      // project_name
      '([^/]*?)/' +
      // bucket_name
      '([^/]*?)/' +
      // folder path + filename
      '(.*)$'
  ).test(path);

// We don't have a lot of content, so let's get a smaller Modal
const customStyles = {
  content: {
    top: '50%',
    left: '50%',
    right: 'auto',
    bottom: 'auto',
    marginRight: '-50%',
    minWidth: '40%',
    transform: 'translate(-50%, -50%)'
  }
};

const FileValue = ({value}) =>
  !value ? (
    <span className='file-missing'> No file </span>
  ) : value instanceof File ? (
    <span className='file-upload'>
      {' '}
      {value.name} ({value.type}){' '}
    </span>
  ) : value === STUB || value.path === STUB ? (
    <span className='file-blank'> Blank file </span>
  ) : value === TEMP || value.path === TEMP ? (
    <span className='file-blank'> {value.original_files[0].name} </span>
  ) : (
    <a href={value.url}> {value.path} </a>
  );

export default function FileAttribute(props) {
  const invoke = useActionInvoker();
  const [metis, setMetis] = useState(false);
  const [error, setError] = useState(false);
  const fileInputRef = useRef(null);

  const {
    metisSelector,
    closeModal,
    selectMetisFile,
    formatFileRevision,
    setTempRevision
  } = useFileActions(metis, error, setMetis, setError, props);

  let {mode, value, revised_value, document, template, attribute} = props;

  if (mode != 'edit')
    return (
      <div className='attribute file'>
        <FileValue value={value} />
      </div>
    );

  let buttons = [
    {
      type: 'upload',
      click: () => fileInputRef.current.click(),
      title: 'Upload a file from your computer'
    },
    {
      type: 'cloud',
      click: () => setMetis(true),
      title: 'Link a file from Metis'
    },
    {
      type: 'stub',
      click: () =>
        invoke(
          reviseDocument(
            document,
            template,
            attribute,
            formatFileRevision(STUB)
          )
        ),
      title: 'Mark this file as blank'
    },
    {
      type: 'remove',
      click: () =>
        invoke(
          reviseDocument(
            document,
            template,
            attribute,
            formatFileRevision(null)
          )
        ),
      title: 'Remove this file link'
    }
  ];

  return (
    <div className='attribute file'>
      <input
        type='file'
        style={{display: 'none'}}
        ref={fileInputRef}
        onChange={setTempRevision}
      />
      {metis && metisSelector()}
      <ButtonBar className='file-buttons' buttons={buttons} />
      <FileValue value={revised_value} />
    </div>
  );

  // const browserState = useReduxState(
  //   browserStateOf({model_name, record_name, tab_name})
  // );
  // const {view, record, tab, revision, template, can_edit} = browserState;
  // const [mode, setMode] = useState('loading');
  // const loading = !view || !template || !record || !tab_name;
  // const {cancelEdits, approveEdits} = useEditActions(setMode, browserState);
  // const {selectOrShowTab, selectDefaultTab, selectTab, showTab} = useTabActions(
  //   browserState,
  //   setMode
  // );
}

// class FileAttribute extends React.Component {
//   constructor(props) {
//     super(props);
//     this.state = {metis: false, error: false};
//   }

//   render() {
//     let {
//       mode,
//       value,
//       revised_value,
//       document,
//       template,
//       attribute,
//       reviseDocument
//     } = this.props;
//     let {metis} = this.state;
//     console.log('revised value', revised_value);
//     if (mode != 'edit')
//       return (
//         <div className='attribute file'>
//           <FileValue value={value} />
//         </div>
//       );

//     let buttons = [
//       {
//         type: 'upload',
//         click: () => this.input.click(),
//         title: 'Upload a file from your computer'
//       },
//       {
//         type: 'cloud',
//         click: () => this.setState({metis: true}),
//         title: 'Link a file from Metis'
//       },
//       {
//         type: 'stub',
//         click: () =>
//           reviseDocument(
//             document,
//             template,
//             attribute,
//             this.formatFileRevision(STUB)
//           ),
//         title: 'Mark this file as blank'
//       },
//       {
//         type: 'remove',
//         click: () =>
//           reviseDocument(
//             document,
//             template,
//             attribute,
//             this.formatFileRevision(null)
//           ),
//         title: 'Remove this file link'
//       }
//     ];

//     return (
//       <div className='attribute file'>
//         <input
//           type='file'
//           style={{display: 'none'}}
//           ref={(input) => (this.input = input)}
//           onChange={this.setTempRevision}
//         />
//         {metis && this.metisSelector()}
//         <ButtonBar className='file-buttons' buttons={buttons} />
//         <FileValue value={revised_value} />
//       </div>
//     );
//   }
// }

function useFileActions(metis, error, setMetis, setError, props) {
  const invoke = useActionInvoker();
  const metisPathRef = useRef(null);

  return {
    metisSelector,
    closeModal,
    selectMetisFile,
    formatFileRevision,
    setTempRevision
  };

  function metisSelector() {
    // TODO: would be nice to make this like a folder / file search
    return (
      <Modal
        isOpen={metis}
        contentLabel='Enter Metis path'
        style={customStyles}
        onRequestClose={closeModal}
        appElement={document.querySelector('#root')}
      >
        <div className='attribute modal file-metis-select'>
          <h2>Enter a Metis path</h2>
          <div className='input-box-wrapper'>
            <label htmlFor='metis-path-input'>Metis path:</label>
            <input
              id='metis-path-input'
              className='full_text metis-path-input'
              type='text'
              ref={metisPathRef}
              placeholder='metis://<project>/<bucket>/<file-path>'
            />
            <div className='modal-button-wrapper'>
              <ButtonBar
                className='modal-buttons'
                buttons={[
                  {type: 'check', click: () => selectMetisFile()},
                  {type: 'cancel', click: () => setMetis(false)}
                ]}
              />
              {error ? (
                <p className='file-metis-error'>Invalid Metis path</p>
              ) : (
                ''
              )}
              <div className='modal-buttons pull-right'>
                <Icon
                  className=''
                  icon='question-circle'
                  title='Help'
                  onClick={() =>
                    window.open(
                      'https://mountetna.github.io/timur.html#managing-data-files',
                      '_blank'
                    )
                  }
                />
              </div>
            </div>
          </div>
        </div>
      </Modal>
    );
  }

  function closeModal() {
    setMetis(false);
  }

  function selectMetisFile() {
    let {document, template, attribute} = props;

    const metisPath = metisPathRef.current.value;

    if (!METIS_PATH_MATCH(metisPath)) {
      setError(true);
      return;
    } else {
      setError(false);
      setMetis(false);
    }

    invoke(
      reviseDocument(
        document,
        template,
        attribute,
        formatFileRevision(metisPath)
      )
    );

    metisPathRef.current.value = null;
  }

  function formatFileRevision(value) {
    return {path: value};
  }

  function setTempRevision(e) {
    e.preventDefault();
    console.log('about to call reviseDocument');
    let {document, template, attribute} = props;
    invoke(
      reviseDocument(
        document,
        template,
        attribute,
        Object.assign(formatFileRevision(TEMP), {
          original_files: e.target.files
        })
      )
    );
  }
}
