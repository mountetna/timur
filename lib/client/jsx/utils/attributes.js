const order = (attribute, identifierFirst) => {
  let {attribute_type} = attribute;
  let map = identifierFirst
    ? {
        identifier: 1,
        parent: 2
      }
    : {
        parent: 1,
        identifier: 2
      };

  map = {
    ...map,
    ...{
      collection: 3,
      table: 4,
      child: 5,
      link: 6,
      file: 7,
      image: 8
    }
  };
  return map[attribute_type] || 9;
};

export const sortAttributes = (attributes, identifierFirst = false) => {
  const sortedAttributes = Object.values(attributes).sort(
    (a, b) =>
      order(a, identifierFirst) - order(b, identifierFirst) ||
      a.attribute_name.localeCompare(b.attribute_name)
  );
  let finalAttributes = {};
  sortedAttributes.forEach((attribute, index) => {
    // inject an index_order for default views
    finalAttributes[attribute.attribute_name] = {
      ...attribute,
      index_order: attribute.index_order ? attribute.index_order : index
    };
  });

  return finalAttributes;
};

export const visibleSortedAttributes = (
  attributes,
  identifierFirst = false
) => {
  return Object.values(sortAttributes(attributes, identifierFirst)).filter(
    (attr) => !attr.hidden
  );
};

export const visibleSortedAttributesWithUpdatedAt = (
  attributes,
  identifierFirst = false
) => {
  return Object.values(sortAttributes(attributes, identifierFirst)).filter(
    (attr) => !attr.hidden || attr.attribute_name === 'updated_at'
  );
};
