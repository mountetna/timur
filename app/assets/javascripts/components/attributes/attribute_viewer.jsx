import React, { Component } from 'react';

import Attribute from './attribute';
import MarkdownAttribute from './markdown_attribute';
import SelectAttribute from './select_attribute';
import ImageAttribute from './image_attribute';
import DocumentAttribute from './document_attribute';
import CheckboxAttribute from './checkbox_attribute';
import DateTimeAttribute from './date_time_attribute';
import TableAttribute from './table_attribute';
import LinkAttribute from './link_attribute';
import { IntegerAttribute, FloatAttribute } from './numeric_attribute';
import CollectionAttribute from './collection_attribute';
import TextAttribute from './text_attribute';
import {MetricsAttributeContainer as MetricsAttribute} from './metrics_attribute';

// The plot components.
import {LinePlotAttributeContainer} from './plot_attributes/line_plot_attribute';
import {BoxPlotAttributeContainer} from './plot_attributes/box_plot_attribute';
import {BarGraphAttributeContainer} from './plot_attributes/bar_graph_attribute';
import {BarPlotAttributeContainer} from './plot_attributes/bar_plot_attribute';
import {StackedBarPlotAttributeContainer} from './plot_attributes/stacked_bar_attribute';
import {SwarmPlotAttributeContainer} from './plot_attributes/swarm_plot_attribute';
import {HistogramAttributeContainer} from './plot_attributes/histogram_attribute';

// The clinical components.
import {DemographicAttributeContainer} from './clinical_attributes/demographic_attribute';
import {DiagnosticAttributeContainer} from './clinical_attributes/diagnostic_attribute';
import {AdverseEventAttributeContainer} from './clinical_attributes/adverse_event_attribute';

export default class AttributeViewer extends Component{
  render(){
    let {attribute} = this.props;
    switch(attribute.attribute_class){

      // Plot components.
      case 'LinePlotAttribute':
        return <LinePlotAttributeContainer {...this.props} />;
      case 'BoxPlotAttribute':
        return <BoxPlotAttributeContainer {...this.props} />;
      case 'BarGraphAttribute':
        return <BarGraphAttributeContainer {...this.props} />;
      case 'BarPlotAttribute':
        return <BarPlotAttributeContainer {...this.props} />;
      case 'StackedBarPlotAttribute':
        return <StackedBarPlotAttributeContainer {...this.props} />;
      case 'SwarmAttribute':
        return <SwarmPlotAttributeContainer {...this.props} />;
      case 'HistogramAttribute':
        return <HistogramAttributeContainer {...this.props} />;

      // Clinical components.
      case 'DemographicAttribute':
        return <DemographicAttributeContainer {...this.props} />;
      case 'DiagnosticAttribute':
        return <DiagnosticAttributeContainer {...this.props} />;
      case 'AdverseEventAttribute':
        return <AdverseEventAttributeContainer {...this.props} />;

      case 'TextAttribute':
        return <TextAttribute {...this.props} />;
      case 'MarkdownAttribute':
        return <MarkdownAttribute {...this.props} />;
      case 'MetricsAttribute':
        return <MetricsAttribute {...this.props} />;
      case 'Magma::CollectionAttribute':
        return <CollectionAttribute {...this.props} />;
      case 'Magma::ForeignKeyAttribute':
      case 'Magma::ChildAttribute':
        return <LinkAttribute {...this.props} />;
      case 'Magma::TableAttribute':
        return <TableAttribute {...this.props} />;
      case 'Magma::FileAttribute':
        return <DocumentAttribute {...this.props} />;
      case 'Magma::ImageAttribute':
        return <ImageAttribute {...this.props} />;
      case 'Magma::Attribute':
        if(attribute.options) return <SelectAttribute {...this.props} />;
        switch(attribute.type){
          case 'TrueClass':
            return <CheckboxAttribute {...this.props} />;
          case 'Integer':
            return <IntegerAttribute {...this.props} />;
          case 'Float':
            return <FloatAttribute {...this.props} />;
          case 'DateTime':
            return <DateTimeAttribute {...this.props} />;
          default:
            return <Attribute {...this.props} />;
        }

      default:
        var msg = 'Could not match attribute '+attribute.name;
        msg += ' with class '+attribute.attribute_class+' to a display class!';
        throw msg;
        return null;
    }
  }
}
