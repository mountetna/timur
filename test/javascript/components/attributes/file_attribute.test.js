import React from 'react';
import {Provider} from 'react-redux';
import {mount, shallow} from 'enzyme';
import {act} from 'react-dom/test-utils';
import {mockStore} from '../../helpers';
import renderer from 'react-test-renderer';
import ReactModal from 'react-modal';
import nock from 'nock';

import ButtonBar from '../../../../lib/client/jsx/components/button_bar';
import {STUB} from '../../../../lib/client/jsx/components/attributes/file_attribute';
import FileAttribute from '../../../../lib/client/jsx/components/attributes/file_attribute';

import ListUpload from 'etna-js/upload/components/list-upload';

describe('FileAttribute', () => {
  let store;

  beforeEach(() => {
    store = mockStore({
      location: {
        path: '/labors/browse/monster/Nemean Lion'
      },
      directory: {
        uploads: {}
      }
    });
  });

  it('renders the button bar while editing', () => {
    const component = mount(
      <Provider store={store}>
        <FileAttribute
          model_name='conquests'
          record_name='Persia'
          template={null}
          value={null}
          mode='edit'
          attribute='gravatar'
          document='Timur'
          revised_value=''
        />
      </Provider>
    );

    const buttons = component.find(ButtonBar);
    expect(buttons.exists()).toBeTruthy();

    const tree = renderer
      .create(
        <Provider store={store}>
          <FileAttribute
            model_name='conquests'
            record_name='Persia'
            template={null}
            value={null}
            mode='edit'
            attribute='gravatar'
            document='Timur'
            revised_value=''
          />
        </Provider>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('does not render action buttons while not editing', () => {
    const value = null;

    const component = mount(
      <Provider store={store}>
        <FileAttribute
          model_name='conquests'
          record_name='Persia'
          template={null}
          value={value}
          mode='view'
          attribute='gravatar'
          document='Timur'
          revised_value=''
        />
      </Provider>
    );

    const buttons = component.find('file-buttons');
    expect(buttons.exists()).toBeFalsy();

    const tree = renderer
      .create(
        <Provider store={store}>
          <FileAttribute
            model_name='conquests'
            record_name='Persia'
            template={null}
            value={value}
            mode='view'
            attribute='gravatar'
            document='Timur'
            revised_value=''
          />
        </Provider>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders blank paths correctly while not editing', () => {
    let value = {path: '::blank'};

    let component = mount(
      <Provider store={store}>
        <FileAttribute
          model_name='conquests'
          record_name='Persia'
          template={null}
          value={value}
          mode='view'
          attribute='gravatar'
          document='Timur'
          revised_value=''
        />
      </Provider>
    );

    expect(component.text().trim()).toEqual('Blank file');

    value = '::blank';

    component = mount(
      <Provider store={store}>
        <FileAttribute
          model_name='conquests'
          record_name='Persia'
          template={null}
          value={value}
          mode='view'
          attribute='gravatar'
          document='Timur'
          revised_value=''
        />
      </Provider>
    );

    expect(component.text().trim()).toEqual('Blank file');

    const tree = renderer
      .create(
        <Provider store={store}>
          <FileAttribute
            model_name='conquests'
            record_name='Persia'
            template={null}
            value={value}
            mode='view'
            attribute='gravatar'
            document='Timur'
            revised_value=''
          />
        </Provider>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders "missing file" correctly while not editing', () => {
    const value = null;

    const component = mount(
      <Provider store={store}>
        <FileAttribute
          model_name='conquests'
          record_name='Persia'
          template={null}
          value={value}
          mode='view'
          attribute='gravatar'
          document='Timur'
          revised_value
        />
      </Provider>
    );

    expect(component.text().trim()).toEqual('No file');

    const tree = renderer
      .create(
        <Provider store={store}>
          <FileAttribute
            model_name='conquests'
            record_name='Persia'
            template={null}
            value={value}
            mode='view'
            attribute='gravatar'
            document='Timur'
            revised_value=''
          />
        </Provider>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders existing files correctly when not editing', () => {
    const value = new File(
      ['Believe me, you are but pismire ant:'],
      'conquest.txt',
      {type: 'text/plain'}
    );

    const component = mount(
      <Provider store={store}>
        <FileAttribute
          model_name='conquests'
          record_name='Persia'
          template={null}
          value={value}
          mode='view'
          attribute='gravatar'
          document='Timur'
          revised_value=''
        />
      </Provider>
    );

    expect(component.text().trim()).toEqual('conquest.txt (text/plain)');

    const tree = renderer
      .create(
        <Provider store={store}>
          <FileAttribute
            model_name='conquests'
            record_name='Persia'
            template={null}
            value={value}
            mode='view'
            attribute='gravatar'
            document='Timur'
            revised_value=''
          />
        </Provider>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders previous value before upload starts', () => {
    store = mockStore({
      location: {
        path: '/labors/browse/monster/Nemean Lion'
      },
      directory: {
        uploads: {}
      }
    });

    const value = {
      path: 'previous-file-value.txt'
    };

    const component = mount(
      <Provider store={store}>
        <FileAttribute
          model_name='conquests'
          record_name='Persia'
          template={null}
          value={value}
          mode='view'
          attribute='gravatar'
          document='Timur'
          revised_value={value}
        />
      </Provider>
    );

    component.setProps({
      value: {
        path:
          'https://metis.test/labors/upload/temporary-path?X-Etna-Signature=foo'
      },
      revised_value: {
        path:
          'https://metis.test/labors/upload/temporary-path?X-Etna-Signature=foo'
      }
    });
    component.update();

    expect(component.text().trim()).toEqual('previous-file-value.txt');

    const uploadControls = component.find(ListUpload);
    expect(uploadControls.exists()).toBeFalsy();
  });

  it('renders upload meter and controls when uploading', () => {
    store = mockStore({
      location: {
        path: '/labors/browse/monster/Nemean Lion'
      },
      directory: {
        uploads: {
          'temporary-file-location': {
            file_name: 'temporary-file-location',
            original_filename: 'stats.txt',
            model_name: 'conquests',
            record_name: 'Persia',
            attribute_name: 'gravatar',
            upload_speeds: [0, 10, 20, 30],
            status: 'active'
          }
        }
      }
    });

    const value = {
      path:
        'https://metis.test/labors/upload/temporary-file-location?X-Etna-Signture=hercules'
    };

    const component = mount(
      <Provider store={store}>
        <FileAttribute
          model_name='conquests'
          record_name='Persia'
          template={null}
          value={value}
          mode='view'
          attribute={{
            attribute_name: 'gravatar'
          }}
          document='Timur'
          revised_value={value}
        />
      </Provider>
    );

    const uploadControls = component.find(ListUpload);
    expect(uploadControls.exists()).toBeTruthy();

    const tree = renderer
      .create(
        <Provider store={store}>
          <FileAttribute
            model_name='conquests'
            record_name='Persia'
            template={null}
            value={value}
            mode='view'
            attribute={{
              attribute_name: 'gravatar'
            }}
            document='Timur'
            revised_value={value}
          />
        </Provider>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders previous file path value if upload cancelled', () => {
    const upload = {
      file_name: 'temporary-file-location',
      original_filename: 'stats.txt',
      model_name: 'conquests',
      record_name: 'Persia',
      attribute_name: 'gravatar',
      upload_speeds: [0, 10, 20, 30],
      status: 'paused'
    };

    let state = {
      location: {
        path: '/labors/browse/monster/Nemean Lion'
      },
      directory: {
        uploads: {
          'temporary-file-location': upload
        }
      }
    };

    store = mockStore(() => state);

    const value = {
      path: 'previous-file-value.txt'
    };

    const component = mount(
      <Provider store={store}>
        <FileAttribute
          model_name='conquests'
          record_name='Persia'
          template={null}
          value={value}
          mode='view'
          attribute={{
            attribute_name: 'gravatar'
          }}
          document='Timur'
          revised_value=''
        />
      </Provider>
    );

    component.setProps({
      value: {
        path:
          'https://metis.test/labors/upload/temporary-file-location?X-Etna-Signature=foo'
      },
      revised_value: {
        path:
          'https://metis.test/labors/upload/temporary-file-location?X-Etna-Signature=foo'
      }
    });
    component.update();

    expect(component.text().trim()).not.toEqual('previous-file-value.txt');
    let uploadControls = component.find(ListUpload);
    expect(uploadControls.exists()).toBeTruthy();

    // Ideally we could test just by clicking on the Cancel button.
    //    But that goes through the service workers, which aren't
    //    really supported in test environments yet?
    // So instead this is a hack, just update store directly.
    state = {
      location: {
        path: '/labors/browse/monster/Nemean Lion'
      },
      directory: {
        uploads: {}
      }
    };
    // Trigger state change.
    act(() => {
      store.dispatch({type: 'CANCEL_UPLOAD', upload});
    });

    component.update();

    expect(component.text().trim()).toEqual('previous-file-value.txt');

    uploadControls = component.find(ListUpload);
    expect(uploadControls.exists()).toBeFalsy();
  });

  it('sends revisions to Magma when upload completes', () => {
    const magmaMock = nock('https://magma.test')
      .post('/update')
      .reply(200, {
        models: {
          conquest: {
            template: {
              attributes: {
                gravatar: {
                  attribute_type: 'image'
                }
              }
            },
            documents: {
              Persia: {
                gravatar: {
                  path: 'https://metis.test/timur/persia/file.txt'
                }
              }
            }
          }
        }
      });

    const uploadUrl =
      'https://metis.test/timur/upload/tmp/temporary-file-location?X-Etna-Signature=foo';

    let state = {
      location: {
        path: '/timur/browse/conquest/Persia'
      },
      directory: {
        uploads: {
          'temporary-file-location': {
            file_name: 'temporary-file-location',
            original_filename: 'stats.txt',
            model_name: 'conquest',
            record_name: 'Persia',
            attribute_name: 'gravatar',
            upload_speeds: [0, 10, 20, 30],
            status: 'complete',
            url: uploadUrl
          }
        }
      }
    };

    store = mockStore(() => state);

    const value = {
      path: uploadUrl
    };

    const component = mount(
      <Provider store={store}>
        <FileAttribute
          model_name='conquest'
          record_name='Persia'
          template={null}
          value={value}
          mode='view'
          attribute={{
            attribute_name: 'gravatar'
          }}
          document='Timur'
          revised_value={null}
        />
      </Provider>
    );

    // Make sure that Magma /update is called
    magmaMock.isDone();
  });

  it('accepts a valid Metis path', () => {
    const component = mount(
      <Provider store={store}>
        <FileAttribute
          model_name='conquests'
          record_name='Persia'
          template={{name: 'Conquests', identifier: 1}}
          value={null}
          mode='edit'
          attribute={{name: 'ExpansionPlans'}}
          document={{'1': 'Timur'}}
          revised_value=''
        />
      </Provider>
    );

    const cloudButton = component.find('.cloud').first();
    cloudButton.simulate('click');

    let modal = component.find(ReactModal);
    const metisPathInput = modal.find('.file-metis-select').find('input');
    metisPathInput.instance().value = 'metis://project/bucket/file_name.txt';

    const checkButton = modal.find('.check').first();
    checkButton.simulate('click');

    // refresh the modal contents
    modal = component.find(ReactModal);
    expect(modal.find('file-metis-error').exists()).toBeFalsy();

    const actions = store.getActions();
    expect(actions).toEqual([
      {
        type: 'REVISE_DOCUMENT',
        model_name: 'Conquests',
        record_name: 'Timur',
        revision: {
          ['ExpansionPlans']: {path: 'metis://project/bucket/file_name.txt'}
        }
      }
    ]);
  });

  it('shows an error message when given an invalid Metis path', () => {
    const component = mount(
      <Provider store={store}>
        <FileAttribute
          model_name='conquests'
          record_name='Persia'
          template={{name: 'Conquests', identifier: 1}}
          value={null}
          mode='edit'
          attribute={{name: 'ExpansionPlans'}}
          document={{'1': 'Timur'}}
          revised_value=''
        />
      </Provider>
    );

    const cloudButton = component.find('.cloud').first();
    cloudButton.simulate('click');

    let modal = component.find(ReactModal);
    const metisPathInput = modal.find('.file-metis-select').find('input');
    metisPathInput.instance().value = 'project/bucket/file_name.txt';

    const checkButton = modal.find('.check').first();
    checkButton.simulate('click');

    // refresh the modal contents
    modal = component.find(ReactModal);

    expect(modal.find('.file-metis-error').exists()).toBeTruthy();

    const actions = store.getActions();
    expect(actions).toEqual([]);
  });

  it('dispatches an action to mark a file as blank', () => {
    const component = mount(
      <Provider store={store}>
        <FileAttribute
          model_name='conquests'
          record_name='Persia'
          template={{name: 'Conquests', identifier: 1}}
          value={null}
          mode='edit'
          attribute={{name: 'ExpansionPlans'}}
          document={{'1': 'Timur'}}
          revised_value=''
        />
      </Provider>
    );

    const stubButton = component.find('.stub').first();
    stubButton.simulate('click');

    const actions = store.getActions();
    expect(actions).toEqual([
      {
        type: 'REVISE_DOCUMENT',
        model_name: 'Conquests',
        record_name: 'Timur',
        revision: {
          ['ExpansionPlans']: {path: STUB}
        }
      }
    ]);
  });

  it('dispatches an action to remove a file', () => {
    const component = mount(
      <Provider store={store}>
        <FileAttribute
          model_name='conquests'
          record_name='Persia'
          template={{name: 'Conquests', identifier: 1}}
          value={null}
          mode='edit'
          attribute={{name: 'ExpansionPlans'}}
          document={{'1': 'Timur'}}
          revised_value=''
        />
      </Provider>
    );

    const removeButton = component.find('.remove').first();
    removeButton.simulate('click');

    const actions = store.getActions();
    expect(actions).toEqual([
      {
        type: 'REVISE_DOCUMENT',
        model_name: 'Conquests',
        record_name: 'Timur',
        revision: {
          ['ExpansionPlans']: {path: null}
        }
      }
    ]);
  });

  // TODO: test for dispatching action when file uploaded
  //   * This has to talk to both Metis and Magma, so a little more
  //      involved.
});
