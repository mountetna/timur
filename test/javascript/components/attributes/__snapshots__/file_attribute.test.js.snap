// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileAttribute does not render action buttons while not editing 1`] = `
<div
  className="attribute file"
>
  <span
    className="file-missing"
  >
     No file 
  </span>
</div>
`;

exports[`FileAttribute renders "missing file" correctly while not editing 1`] = `
<div
  className="attribute file"
>
  <span
    className="file-missing"
  >
     No file 
  </span>
</div>
`;

exports[`FileAttribute renders blank paths correctly while not editing 1`] = `
<div
  className="attribute file"
>
  <span
    className="file-blank"
  >
     Blank file 
  </span>
</div>
`;

exports[`FileAttribute renders existing files correctly when not editing 1`] = `
<div
  className="attribute file"
>
  <span
    className="file-upload"
  >
     
    conquest.txt
     (
    text/plain
    )
     
  </span>
</div>
`;

exports[`FileAttribute renders the button bar while editing 1`] = `
<div
  className="attribute file"
>
  <input
    onChange={[Function]}
    style={
      Object {
        "display": "none",
      }
    }
    type="file"
  />
  <div
    className="file-buttons"
  >
    <span
      className="fa-stack fa-fw"
      onClick={[Function]}
      title="Upload a file from your computer"
    >
      <i
        className="icon upload upload fa fa-2x fa-upload"
      />
    </span>
    <span
      className="fa-stack fa-fw"
      onClick={[Function]}
      title="Link a file from Metis"
    >
      <i
        className="icon cloud cloud fa fa-2x fa-cloud"
      />
    </span>
    <span
      className="fa-stack fa-fw"
      onClick={[Function]}
      title="Mark this file as blank"
    >
      <i
        className="icon stub meh-blank fa fa-2x fa-meh-blank"
      />
    </span>
    <span
      className="fa-stack fa-fw"
      onClick={[Function]}
      title="Remove this file link"
    >
      <i
        className="icon remove trash-alt fa fa-2x fa-trash-alt"
      />
    </span>
  </div>
  <span
    className="file-missing"
  >
     No file 
  </span>
</div>
`;

exports[`FileAttribute renders upload meter and controls when uploading 1`] = `
<div
  className="attribute file"
>
  <div
    className="list-entry-group upload"
  >
    <div
      className="list-entry-column-group type"
      style={
        Object {
          "flexBasis": "60px",
        }
      }
    >
      <span
        className="icon-grp fa-stack fa-fw"
      >
        <i
          className="icon undefined upload fa fa-2x fa-upload"
        />
      </span>
    </div>
    <div
      className="list-entry-column-group name"
      style={
        Object {
          "flexBasis": "60%",
        }
      }
    >
      <span
        className="list-entry-file-name"
      >
        stats.txt
      </span>
    </div>
    <div
      className="list-entry-column-group status"
      style={
        Object {
          "flexBasis": "90px",
        }
      }
    />
    <div
      className="list-entry-column-group updated"
      style={
        Object {
          "flexBasis": "30%",
        }
      }
    >
      <div
        className="upload-meter-group"
      >
        <div
          className="upload-meter-tray"
        >
          <div
            className="upload-meter-bar"
            style={
              Object {
                "width": "NaN%",
              }
            }
          />
        </div>
        <div
          className="upload-meter-status"
        >
          <span
            className="icon-grp fa-stack fa-fw"
          >
            <i
              className="icon active upload fa fa-2x fa-upload"
            />
          </span>
        </div>
        <div
          className="upload-meter-completion"
        >
          <span
            className="completed"
            title="kilobytes uploaded"
          >
            NaN undefinediB
          </span>
           of 
          NaN undefinediB
        </div>
        <div
          className="upload-meter-speed"
        >
          <span
            title="kilobits per second"
          >
            117.2 Kbps
          </span>
        </div>
      </div>
    </div>
    <div
      className="list-entry-column-group size"
      style={
        Object {
          "flexBasis": "10%",
        }
      }
    />
    <div
      className="list-entry-column-group control"
      style={
        Object {
          "flexBasis": "100px",
        }
      }
    >
      <div
        className="upload-control-group"
      >
        <button
          className="upload-control-btn"
          onClick={[Function]}
        >
          <span
            className="fas fa-fw fa-pause"
          />
        </button>
        <button
          className="upload-control-btn"
          onClick={[Function]}
        >
          <span
            className="fas fa-fw fa-times"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;
